File: Tests/Parsing/FormScrapingTests.m
//  FormScrapingTests.m
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "AwfulParsingTestCase.h"
#import "AwfulForm.h"
#import "AwfulFormScraper.h"

@interface FormScrapingTests : AwfulParsingTestCase

@end

@implementation FormScrapingTests

+ (Class)scraperClass
{
    return [AwfulFormScraper class];
}

- (void)testReply
{
    NSArray *forms = [self scrapeFixtureNamed:@"newreply"];
    XCTAssertEqual(forms.count, (NSUInteger)1);
    AwfulForm *form = forms[0];
    XCTAssertEqual(form.threadTags.count, (NSUInteger)0);
    NSMutableDictionary *parameters = [form recommendedParameters];
    XCTAssertEqualObjects(parameters[@"action"], @"postreply");
    XCTAssertEqualObjects(parameters[@"threadid"], @"3507451");
    XCTAssertEqualObjects(parameters[@"formkey"], @"0253d85a945b60daa0165f718df82b8a");
    XCTAssertEqualObjects(parameters[@"form_cookie"], @"80c74b48f557");
    XCTAssertNotNil(parameters[@"message"]);
    XCTAssertNotEqual([parameters[@"message"] rangeOfString:@"terrible"].location, NSNotFound);
    XCTAssertNotNil(parameters[@"parseurl"]);
    XCTAssertNotNil(parameters[@"bookmark"]);
    XCTAssertNil(parameters[@"disablesmilies"]);
    XCTAssertNil(parameters[@"signature"]);
}

- (void)testThread
{
    NSArray *forms = [self scrapeFixtureNamed:@"newthread"];
    XCTAssertEqual(forms.count, (NSUInteger)1);
    AwfulForm *form = forms[0];
    XCTAssertEqual(form.threadTags.count, (NSUInteger)51);
    NSArray *allThreadTags = [AwfulThreadTag fetchAllInManagedObjectContext:self.managedObjectContext];
    XCTAssertEqual(allThreadTags.count, form.threadTags.count);
    XCTAssertNil(form.secondaryThreadTags);
    NSArray *textNames = [form.texts valueForKey:@"name"];
    XCTAssertTrue([textNames containsObject:@"subject"]);
    XCTAssertTrue([textNames containsObject:@"message"]);
    NSMutableDictionary *parameters = [form recommendedParameters];
    XCTAssertEqualObjects(parameters[@"forumid"], @"1");
    XCTAssertEqualObjects(parameters[@"action"], @"postthread");
    XCTAssertEqualObjects(parameters[@"formkey"], @"0253d85a945b60daa0165f718df82b8a");
    XCTAssertEqualObjects(parameters[@"form_cookie"], @"e29a15add831");
    XCTAssertEqualObjects(parameters[@"parseurl"], @"yes");
    XCTAssertEqualObjects(parameters[@"bookmark"], @"yes");
}

- (void)testAskTellThread
{
    NSArray *forms = [self scrapeFixtureNamed:@"newthread-at"];
    XCTAssertEqual(forms.count, (NSUInteger)1);
    AwfulForm *form = forms[0];
    XCTAssertEqual(form.threadTags.count, (NSUInteger)55);
    XCTAssertEqual(form.secondaryThreadTags.count, (NSUInteger)2);
    NSArray *secondaryTags = [AwfulThreadTag fetchAllInManagedObjectContext:self.managedObjectContext
                                                    matchingPredicateFormat:@"imageName IN { 'ama', 'tma' }"];
    XCTAssertEqual(secondaryTags.count, form.secondaryThreadTags.count);
}

- (void)testSAMartThread
{
    NSArray *forms = [self scrapeFixtureNamed:@"newthread-samart"];
    XCTAssertEqual(forms.count, (NSUInteger)1);
    AwfulForm *form = forms[0];
    XCTAssertEqual(form.threadTags.count, (NSUInteger)69);
    XCTAssertEqual(form.secondaryThreadTags.count, (NSUInteger)4);
    NSArray *secondaryTags = [AwfulThreadTag fetchAllInManagedObjectContext:self.managedObjectContext
                                                    matchingPredicateFormat:@"imageName LIKE 'icon*ing'"];
    XCTAssertEqual(secondaryTags.count, form.secondaryThreadTags.count);
}

- (void)testMessage
{
    NSArray *forms = [self scrapeFixtureNamed:@"private-reply"];
    XCTAssertEqual(forms.count, (NSUInteger)1);
    AwfulForm *form = forms[0];
    AwfulFormItem *messageItem;
    for (AwfulFormItem *text in form.texts) {
        if ([text.name isEqualToString:@"message"]) {
            messageItem = text;
            break;
        }
    }
    XCTAssertNotNil(messageItem);
    XCTAssertNotEqual([messageItem.value rangeOfString:@"InFlames235 wrote"].location, NSNotFound);
}

@end

File: Core/Tests/FormScrapingTests.swift
//  FormScrapingTests.swift
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

@testable import AwfulCore
import CoreData
import XCTest

final class FormScrapingTests: XCTestCase {
    private lazy var storeCoordinator: NSPersistentStoreCoordinator = {
        let modelURL = Bundle(for: AwfulManagedObject.self).url(forResource: "Awful", withExtension: "momd")!
        let model = NSManagedObjectModel(contentsOf: modelURL)
        return NSPersistentStoreCoordinator(managedObjectModel: model)
    }()
    private var memoryStore: NSPersistentStore!
    private var managedObjectContext: NSManagedObjectContext!

    override func setUp() {
        super.setUp()

        memoryStore = try! storeCoordinator.addPersistentStore(ofType: NSInMemoryStoreType, configurationName: nil, at: nil, options: nil)

        managedObjectContext = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
        managedObjectContext.persistentStoreCoordinator = storeCoordinator
    }

    override func tearDown() {
        managedObjectContext = nil

        try! storeCoordinator.remove(memoryStore)
        memoryStore = nil

        super.tearDown()
    }

    fileprivate func scrapeFormFixtureNamed(_ fixtureName: String) -> AwfulForm {
        let document = fixtureNamed(fixtureName)
        let formElement = document.firstNode(matchingSelector: "form[name='vbform']")
        let form = AwfulForm(element: formElement!)
        form?.scrapeThreadTags(into: managedObjectContext)
        do {
            try managedObjectContext.save()
        }
        catch {
            fatalError("error saving context after scraping thread tags: \(error)")
        }
        
        return form!
    }

    func testReply() {
        let form = scrapeFormFixtureNamed("newreply")
        XCTAssertTrue(form.threadTags == nil)
        let parameters = form.recommendedParameters()
        XCTAssertEqual((parameters?["action"] as! String), "postreply")
        XCTAssertEqual((parameters?["threadid"] as! String), "3507451")
        XCTAssertEqual((parameters?["formkey"] as! String), "0253d85a945b60daa0165f718df82b8a")
        XCTAssertEqual((parameters?["form_cookie"] as! String), "80c74b48f557")
        XCTAssertTrue((parameters?["message"] as! String).range(of: "terrible") != nil)
        XCTAssertNotNil(parameters?["parseurl"])
        XCTAssertNotNil(parameters?["bookmark"])
        XCTAssert(parameters?["disablesmilies"] == nil)
        XCTAssert(parameters?["signature"] == nil)
    }

    func testReplyWithAmazonSearch() {
        let form = scrapeFormFixtureNamed("newreply-amazon-form")
        XCTAssertNotNil(form.recommendedParameters()?["threadid"])
    }

    func testThread() {
        let form = scrapeFormFixtureNamed("newthread")
        XCTAssertTrue(form.threadTags?.count == 51)
        XCTAssertEqual(fetchAll(ThreadTag.self, inContext: managedObjectContext).count, form.threadTags?.count)
        XCTAssertTrue(form.secondaryThreadTags == nil)
        let parameters = form.allParameters
        XCTAssertNotNil(parameters?["subject"])
        XCTAssertNotNil(parameters?["message"])
        XCTAssertEqual(parameters?["forumid"], "1")
        XCTAssertEqual(parameters?["action"], "postthread")
        XCTAssertEqual(parameters?["formkey"], "0253d85a945b60daa0165f718df82b8a")
        XCTAssertEqual(parameters?["form_cookie"], "e29a15add831")
        XCTAssertEqual(parameters?["parseurl"], "yes")
        XCTAssertEqual(parameters?["bookmark"], "yes")
    }
    
    func testAskTellThread() {
        let form = scrapeFormFixtureNamed("newthread-at")
        XCTAssertTrue(form.threadTags?.count == 55)
        XCTAssertTrue(form.secondaryThreadTags?.count == 2)
        let secondaryTags = fetchAll(ThreadTag.self, inContext: managedObjectContext, matchingPredicate: NSPredicate(format: "imageName IN { 'ama', 'tma' }"))
        XCTAssertEqual(secondaryTags.count, form.secondaryThreadTags?.count);
    }
    
    func testSAMartThread() {
        let form = scrapeFormFixtureNamed("newthread-samart")
        XCTAssertTrue(form.threadTags?.count == 69)
        XCTAssertTrue(form.secondaryThreadTags?.count == 4)
        let possibleSecondaryTags = fetchAll(ThreadTag.self, inContext: managedObjectContext, matchingPredicate: NSPredicate(format: "imageName LIKE 'icon*ing'"))
        let secondaryTags = possibleSecondaryTags.filter { Int($0.threadTagID!)! < 5 }
        XCTAssertEqual(secondaryTags.count, form.secondaryThreadTags?.count)
    }
    
    func testMessage() {
        let form = scrapeFormFixtureNamed("private-reply")
        let parameters = form.allParameters
        XCTAssertNotNil(parameters?["message"])
        XCTAssertTrue(parameters?["message"]?.range(of: "InFlames235 wrote") != nil)
    }
}
